airbyte-cdk/python/airbyte_cdk/entrypoint.py

```diff
@@ -172,6 +186,60 @@ def launch(source: Source, args: List[str]):
         print(message)
 
 
+def _init_internal_request_filter():
+    """
+    Wraps the Python requests library to prevent sending requests to internal URL endpoints.
+    """
+    wrapped_fn = Session.send
+
+    @wraps(wrapped_fn)
+    def filtered_send(self, request, **kwargs):
+        parsed_url = urlparse(request.url)
+
+        if parsed_url.scheme not in VALID_URL_SCHEMES:
+            raise ValueError(
+                "Invalid Protocol Scheme: The endpoint that data is being requested from is using an invalid or insecure "
+                + f"protocol {parsed_url.scheme}. Valid protocol schemes: {','.join(VALID_URL_SCHEMES)}"
+            )
+
+        if not parsed_url.hostname:
+            raise ValueError("Invalid URL specified: The endpoint that data is being requested from is not a valid URL")
+
+        try:
+            is_private = _is_private_url(parsed_url.hostname, parsed_url.port)
+            if is_private:
+                raise ValueError(
+                    "Invalid URL endpoint: The endpoint that data is being requested from belongs to a private network. Source "
+                    + "connectors only support requesting data from public API endpoints."
+                )
+        except socket.gaierror:
+            # This is a special case where the developer specifies an IP address string that is not formatted correctly like trailing
+            # whitespace which will fail the socket IP lookup. This only happens when using IP addresses and not text hostnames.
+            raise ValueError(f"Invalid hostname or IP address '{parsed_url.hostname}' specified.")
+
+        return wrapped_fn(self, request, **kwargs)
+
+    Session.send = filtered_send
+
+
+@cache
+def _is_private_url(hostname: str, port: int) -> bool:
+    """
+    Helper method that checks if any IPs are on a private network. We specifically only pass hostname and port
+    parameters because it will maximize the hit rate across different request permutations and keep the size of
+    of cache lower since the range of hostname and port should be small on a per-sync basis.
```

Comment:
emphasizing that rationale for why hostname and port as parameters is important instead of the parsed_url which contains other things that could reduce our hit rate, like path, params, etc